server/session/handler_item_stack_request.go

```diff
@@ -183,6 +190,135 @@ func call(ctx *event.Context, slot int, it item.Stack, f func(ctx *event.Context
 	return nil
 }
 
+// handleCraft handles the CraftRecipe request action.
+func (h *ItemStackRequestHandler) handleCraft(a *protocol.CraftRecipeStackRequestAction, s *Session) error {
+	craft, ok := s.recipes[a.RecipeNetworkID]
+	if !ok {
+		return fmt.Errorf("recipe with network id %v does not exist", a.RecipeNetworkID)
+	}
+	_, shaped := craft.(recipe.ShapedRecipe)
+	_, shapeless := craft.(recipe.ShapelessRecipe)
+	if !shaped && !shapeless {
+		return fmt.Errorf("recipe with network id %v is not a shaped or shapeless recipe", a.RecipeNetworkID)
+	}
+
+	size := s.craftingSize()
+	offset := s.craftingOffset()
+	consumed := make([]bool, size)
+	for _, expected := range craft.Input() {
+		var processed bool
+		for slot := offset; slot < offset+size; slot++ {
+			if consumed[slot-offset] {
+				// We've already consumed this slot, skip it.
+				continue
+			}
+			has, _ := s.ui.Item(int(slot))
+			if has.Empty() != expected.Empty() || has.Count() < expected.Count() || !has.Comparable(expected) {
+				// We can't process this item, as it's not a part of the recipe.
+				continue
+			}
+			processed, consumed[slot-offset] = true, true
+			st := has.Grow(-expected.Count())
+			h.setItemInSlot(protocol.StackRequestSlotInfo{
+				ContainerID:    containerCraftingGrid,
+				Slot:           byte(slot),
+				StackNetworkID: item_id(st),
+			}, st, s)
+			break
+		}
+		if !processed {
+			return fmt.Errorf("recipe %v: could not consume expected item: %v", a.RecipeNetworkID, expected)
+		}
+	}
+
+	output := craft.Output()
+	h.setItemInSlot(protocol.StackRequestSlotInfo{
+		ContainerID:    containerCraftingGrid,
+		Slot:           craftingGridResult,
+		StackNetworkID: item_id(output[0]),
+	}, output[0], s)
+	return nil
+}
+
+// handleAutoCraft handles the AutoCraftRecipe request action.
+func (h *ItemStackRequestHandler) handleAutoCraft(a *protocol.AutoCraftRecipeStackRequestAction, s *Session) error {
+	craft, ok := s.recipes[a.RecipeNetworkID]
+	if !ok {
+		return fmt.Errorf("recipe with network id %v does not exist", a.RecipeNetworkID)
+	}
+	_, shaped := craft.(recipe.ShapedRecipe)
+	_, shapeless := craft.(recipe.ShapelessRecipe)
+	if !shaped && !shapeless {
+		return fmt.Errorf("recipe with network id %v is not a shaped or shapeless recipe", a.RecipeNetworkID)
+	}
+
+	input := make([]item.Stack, 0, len(craft.Input()))
+	for _, i := range craft.Input() {
+		input = append(input, i.Grow(i.Count()*(int(a.TimesCrafted)-1)))
+	}
+
+	expectancies := make([]item.Stack, 0, len(input))
+	for _, i := range input {
+		if i.Empty() {
+			// We don't actually need this item - it's empty, so avoid putting it in our expectancies.
+			continue
+		}
+
+		if ind := slices.IndexFunc(expectancies, func(stack item.Stack) bool {
+			return stack.Comparable(i)
```

Comment:
If you change this to `i.Comparable`, you can simply do `slices.IndexFunc(expectancies, i.Comparable)`.